### PR TITLE
UCP/KEEPALIVE: add AM-based keepalive

### PR DESCRIPTION
**Pull request by hoopoepg**
- for connect-to-iface EPs use AM based keepalive
- new way is implemented using ping-pong pattern

### CODE REVIEW

**Comment by hoopoepg** [comment]
bot:pipe:retest

File: src/ucp/core/ucp_ep.h [hunk 1]
```diff
@@ -514,6 +514,9 @@ ucs_status_ptr_t ucp_ep_flush_internal(ucp_ep_h ep, unsigned req_flags,
                                        ucp_request_callback_t flushed_cb,
                                        const char *debug_name);
 
+ucs_status_ptr_t ucp_ep_close_flushed(ucp_ep_h ep,
+                                      const ucp_request_param_t *param) ;
```
Comment (brminich) [inline src/ucp/core/ucp_ep.h:518]: extra space before `;`

Comment (brminich) [inline src/ucp/core/ucp_ep.h:518]: extra space before `;`

File: src/ucp/wireup/wireup.c [hunk 2]
```diff
@@ -602,6 +606,61 @@ ucp_wireup_process_reply(ucp_worker_h worker, ucp_ep_h ep,
...
+out:
+    return;
+
+failed_destroy_ep:
+    ucp_ep_destroy_internal(reply_ep);
+    goto out;
+}
+
+static UCS_F_NOINLINE void
+ucp_wireup_process_ep_failed(ucp_worker_h worker, ucp_ep_h ep)
```
Comment (brminich) [inline src/ucp/wireup/wireup.c:658]: why need this wrapper? Can just call `ucp_worker_set_ep_failed`

File: src/ucp/core/ucp_ep.c [hunk 3]
```diff
@@ -944,12 +944,39 @@ void ucp_ep_discard_lanes(ucp_ep_h ep, ucs_status_t status)
     }
 }
 
-ucs_status_ptr_t ucp_ep_close_nbx(ucp_ep_h ep, const ucp_request_param_t *param)
+
```
Comment (brminich) [inline src/ucp/core/ucp_ep.c:947]: extra line

**Comment by yosefe** [comment]
bot:pipe:retest

File: src/ucp/core/ucp_ep.c (hunk 3 above)
Comment (hoopoepg) [reply src/ucp/core/ucp_ep.c]: removed

File: src/ucp/core/ucp_ep.h (hunk 1 above)
Comment (hoopoepg) [reply src/ucp/core/ucp_ep.h]: yep, removed

Comment (hoopoepg) [reply src/ucp/core/ucp_ep.h]: removed too

File: src/ucp/wireup/wireup.c (hunk 2 above)
Comment (hoopoepg) [reply src/ucp/wireup/wireup.c]: the general way to process wireup messages... removed

File: src/ucp/core/ucp_ep.c [hunk 4]
```diff
@@ -944,12 +944,38 @@ void ucp_ep_discard_lanes(ucp_ep_h ep, ucs_status_t status)
...
+                                      const ucp_request_param_t *param)
 {
-    ucp_worker_h  worker = ep->worker;
-    void          *request = NULL;
+    void *request;
     ucp_request_t *close_req;
 
+    request = ucp_ep_flush_internal(ep, 0, param, NULL,
+                                    ucp_ep_close_flushed_callback, "close");
+    if (!UCS_PTR_IS_PTR(request)) {
```
Comment (brminich) [inline src/ucp/core/ucp_ep.c:955]: minor:
better do `if (UCS_PTR_IS_PTR(request)) return request;`

File: src/ucp/core/ucp_ep.c [hunk 5]
```diff
@@ -2319,15 +2329,31 @@ void ucp_ep_do_keepalive(ucp_ep_h ep, ucp_lane_map_t *lane_map)
...
-            *lane_map &= ~UCS_BIT(lane);
-        } else if (status != UCS_ERR_NO_RESOURCE) {
-            ucs_warn("unexpected return status from uct_ep_check(ep=%p): %s",
-                     ep, ucs_status_string(status));
+        /* in case if remote ID is defined, wireup is completed and EP is
+         * in connect-to-iface mode then use UCP/AM keepalive */
+        if ((ep->flags & UCP_EP_FLAG_REMOTE_ID) && /* we have remote ID to check */
+            (ep->flags & UCP_EP_FLAG_REMOTE_CONNECTED) && /* wireup completed */
+            (ucp_ep_get_iface_attr(ep, lane)->cap.flags &
+             UCT_IFACE_FLAG_CONNECT_TO_IFACE)) { /* connect-to-iface */
```
Comment (brminich) [inline src/ucp/core/ucp_ep.c:2343]: maybe better have two different maps to avoid these check for every lane?

File: src/ucp/core/ucp_ep.c (hunk 4 above)
Comment (hoopoepg) [reply src/ucp/core/ucp_ep.c]: ok, done

File: src/ucp/core/ucp_ep.c (hunk 5 above)
Comment (hoopoepg) [reply src/ucp/core/ucp_ep.c:2345]: actually same UCT EP may be processed as ep_check & AM ping-pong depends from EP state (connected or not)

**Comment by hoopoepg** [comment]
depends from https://github.com/openucx/ucx/pull/6197

**Comment by hoopoepg** [comment]
resolved conflicts, have to squash changes

**Comment by hoopoepg** [comment]
bot:pipe:retest

**Comment by hoopoepg** [comment]
@brminich more comments?

File: src/ucp/wireup/wireup.c [hunk 6]
```diff
@@ -606,6 +610,54 @@ ucp_wireup_process_reply(ucp_worker_h worker, ucp_ep_h ep,
...
+ucp_wireup_process_ep_check(ucp_worker_h worker, ucp_ep_h ep,
+                            const ucp_wireup_msg_t *msg,
+                            const ucp_unpacked_address_t *remote_address)
+{
+    ucs_status_t status;
+    ucp_ep_h reply_ep;
+    unsigned addr_indices[UCP_MAX_LANES];
+    ucs_status_ptr_t req;
+
+    if (ep != NULL) {
```
Comment (brminich) [inline src/ucp/wireup/wireup.c:623]: do we need to check for `FAILED|CLOSED` flags?

Comment (hoopoepg) [reply src/ucp/wireup/wireup.c:623]: not needed - failed & closed EP's will be destroyed later and will be processed on next ep_check iteration

**Comment by hoopoepg** [comment]
@brminich ok to approve?

**Review by brminich — APPROVED** [review]

**Comment by hoopoepg** [comment]
@yosefe ok to merge?

File: src/ucp/wireup/wireup.c (hunk 6 above)
Comment (yosefe) [reply src/ucp/wireup/wireup.c:623]: pls add comment: if endpoint exists - do not sent a reply

File: src/ucp/wireup/wireup.c [hunk 7]
```diff
@@ -606,6 +610,54 @@ ucp_wireup_process_reply(ucp_worker_h worker, ucp_ep_h ep,
...
+    ucp_ep_update_remote_id(reply_ep, msg->src_ep_id);
+    status = ucp_wireup_msg_send(reply_ep, UCP_WIREUP_MSG_EP_FAILED,
+                                 0, NULL);
+    if (status != UCS_OK) {
+        goto failed_destroy_ep;
+    }
+
+    req = ucp_ep_close_flushed(reply_ep, &ucp_request_null_param);
+    if (UCS_PTR_IS_PTR(req)) {
+        ucp_request_free(req);
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:650]: 1. goto out_req_put
2. use ucp_request_put

File: src/ucp/wireup/wireup.c [hunk 8]
```diff
@@ -606,6 +610,54 @@ ucp_wireup_process_reply(ucp_worker_h worker, ucp_ep_h ep,
...
+    }
+
+out:
+    return;
+
+failed_destroy_ep:
+    ucp_ep_destroy_internal(reply_ep);
+    goto out;
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:658]: no need for goto out here

**Review by brminich — APPROVED** [review]

File: src/ucp/core/ucp_ep.c [hunk 9]
```diff
@@ -2382,6 +2382,15 @@ int ucp_ep_config_test_rndv_support(const ucp_ep_config_t *config)
            (config->key.cm_lane  != UCP_NULL_LANE);
 }
 
+static UCS_F_ALWAYS_INLINE int
+ucp_ep_is_am_keepalive(ucp_ep_h ep, ucp_lane_index_t lane)
+{
+    return (ep->flags & UCP_EP_FLAG_REMOTE_ID) && /* remote ID defined */
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:2388]: pls add a comment that if we have 2iface transport we need to send an active-message based keepalive to check the remote endpoint still exists

File: src/ucp/core/ucp_ep.c [hunk 10]
```diff
@@ -2412,8 +2419,9 @@ void ucp_ep_do_keepalive(ucp_ep_h ep, ucp_lane_map_t *lane_map)
             if (status == UCS_OK) {
                 *lane_map &= ~UCS_BIT(lane);
             } else if (status != UCS_ERR_NO_RESOURCE) {
-                ucs_warn("unexpected return status from uct_ep_check(ep=%p): %s",
-                         ep, ucs_status_string(status));
+                ucs_warn(
+                        "unexpected return status from uct_ep_check(ep=%p): %s",
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:2423]: maybe just "uct_ep_check(ep=%p) failed: %s"

File: src/ucp/wireup/wireup.c [hunk 11]
```diff
@@ -64,7 +64,7 @@ static const char* ucp_wireup_msg_str(uint8_t msg_type)
         return "ACK";
     case UCP_WIREUP_MSG_EP_CHECK:
         return "EP_CHECK";
-    case UCP_WIREUP_MSG_EP_FAILED:
+    case UCP_WIREUP_MSG_EP_REMOVED:
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:67]: pls change returned string as well

File: src/ucp/wireup/wireup.c [hunk 12]
```diff
@@ -653,23 +655,23 @@ ucp_wireup_process_ep_check(ucp_worker_h worker, ucp_ep_h ep,
...
 
-    req = ucp_ep_close_flushed(reply_ep, &ucp_request_null_param);
-    if (UCS_PTR_IS_PTR(req)) {
-        ucp_request_free(req);
+    req = ucp_ep_flush_and_close(reply_ep, &ucp_request_null_param);
+    if (!UCS_PTR_IS_PTR(req)) {
+        goto out;
     }
 
+    ucp_request_release(req);
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:668]: maybe create the request with flag RELEASED ?

File: src/ucp/core/ucp_ep.c [hunk 13]
```diff
@@ -944,12 +944,40 @@ void ucp_ep_discard_lanes(ucp_ep_h ep, ucs_status_t status)
...
+    void *request;
     ucp_request_t *close_req;
 
+    request = ucp_ep_flush_internal(ep, 0, param, NULL,
+                                    ucp_ep_close_flushed_callback, "close");
+    if (UCS_PTR_IS_PTR(request)) {
+        return request;
+    }
+
+    if (ucp_ep_is_cm_local_connected(ep)) {
```
[the remark this answers is not in the document]
Comment (yosefe) [reply src/ucp/core/ucp_ep.c:959]: then why we need this common function?
our new flow could just call ucp_ep_flush_internal() with flag RELEASED

File: src/ucp/wireup/wireup.c [hunk 14]
```diff
@@ -676,6 +731,13 @@ static ucs_status_t ucp_wireup_msg_handler(void *arg, void *data,
         ucp_wireup_process_request(worker, ep, msg, &remote_address);
     } else if (msg->type == UCP_WIREUP_MSG_REPLY) {
         ucp_wireup_process_reply(worker, ep, msg, &remote_address);
+    } else if (msg->type == UCP_WIREUP_MSG_EP_CHECK) {
+        ucs_assert(msg->dst_ep_id != UCP_EP_ID_INVALID);
```
Comment (dmitrygx) [inline src/ucp/wireup/wireup.c:735]: can we remove the assert and use `UCP_WIREUP_MSG_CHECK(msg, ep, UCP_WIREUP_MSG_EP_CHECK);` inside the handler?

File: src/ucp/core/ucp_ep.c [hunk 15]
```diff
@@ -979,21 +979,21 @@ ucs_status_ptr_t ucp_ep_close_nbx(ucp_ep_h ep, const ucp_request_param_t *param)
...
+                close_req = ucp_ep_cm_close_request_get(ep, param); 
+                if (close_req != NULL) {    
+                    request = close_req + 1;    
+                    ucp_ep_set_close_request(ep, close_req, "close");   
+                } else {    
+                    request = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);    
+                }   
+            } else {    
+                ucp_ep_disconnected(ep, 0); 
+            }
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:996]: extra trailing spaces

File: src/ucp/core/ucp_ep.c [hunk 16]
```diff
@@ -947,7 +947,7 @@ void ucp_ep_discard_lanes(ucp_ep_h ep, ucs_status_t status)
 ucs_status_ptr_t ucp_ep_close_nbx(ucp_ep_h ep, const ucp_request_param_t *param)
 {
     ucp_worker_h  worker = ep->worker;
-    void          *request = NULL;
+    void          *request = NULL;
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:950]: extra space

File: src/ucp/core/ucp_ep.c [hunk 17]
```diff
@@ -2370,21 +2370,48 @@ int ucp_ep_config_test_rndv_support(const ucp_ep_config_t *config)
...
         ucs_assert(lane < UCP_MAX_LANES);
-        /* coverity[overrun-local] */
-        status = uct_ep_check(ep->uct_eps[lane], 0, NULL);
-        if (status == UCS_OK) {
+        /* in case if remote ID is defined, wireup is completed and EP is
+         * in connect-to-iface mode then use UCP/AM keepalive */
+        if (ucp_ep_is_am_keepalive(ep, lane)) {
+            status = ucp_wireup_msg_send(ep, UCP_WIREUP_MSG_EP_CHECK,
+                                         UCS_BIT(ucp_ep_get_rsc_index(ep,
+                                                                      lane)),
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:2402]: minor: add helper var rsc_index

File: src/ucp/core/ucp_ep.h [hunk 18]
```diff
@@ -516,6 +516,9 @@ ucs_status_ptr_t ucp_ep_flush_internal(ucp_ep_h ep, unsigned req_flags,
                                        ucp_request_callback_t flushed_cb,
                                        const char *debug_name);
 
+ucs_status_ptr_t
+ucp_ep_flush_and_close(ucp_ep_h ep, const ucp_request_param_t *param);
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.h:520]: remove

File: src/ucp/wireup/wireup.c [hunk 19]
```diff
@@ -620,6 +623,73 @@ ucp_wireup_process_reply(ucp_worker_h worker, ucp_ep_h ep,
     }
 }
 
+static void ucp_wireup_close_flushed_callback(ucp_request_t *req)
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:626]: duplicate of the code in ucp_ep.c?

File: src/ucp/wireup/wireup.c [hunk 20]
```diff
@@ -620,6 +623,73 @@ ucp_wireup_process_reply(ucp_worker_h worker, ucp_ep_h ep,
...
+        return;
+    }
+
+    /* if endpoint does not exist - create a temporary endpoint to send a
+     * UCP_WIREUP_MSG_EP_REMOVED reply */
+    status = ucp_worker_create_ep(worker, 0, remote_address->name,
+                                  "wireup ep_check reply", &reply_ep);
+    if (status != UCS_OK) {
+        ucs_error("failed to create EP: %s", ucs_status_string(status));
+        goto out;
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:663]: since "out:" label does not do anything now, can replace by "return"

File: src/ucp/wireup/wireup_cm.c [hunk 21]
```diff
@@ -629,7 +629,7 @@ static void ucp_ep_cm_remote_disconnect_progress(ucp_ep_h ucp_ep)
     if (ucp_ep->flags & UCP_EP_FLAG_CLOSED) {
         /* the ep is remote connected (checked above) and closed by API but
          * close req is not valid yet (checked above), it will be set later
-         * from scheduled @ref ucp_ep_close_flushed_callback */
+         * from scheduled @ref ucp_ep_flush_and_close_callback */
```
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:632]: revert?

File: src/ucp/core/ucp_ep.c (hunk 16 above)
Comment (hoopoepg) [reply src/ucp/core/ucp_ep.c]: yep, copy-paste issue

File: src/ucp/wireup/wireup_cm.c (hunk 21 above)
Comment (hoopoepg) [reply src/ucp/wireup/wireup_cm.c]: reverted

File: src/ucp/core/ucp_ep.c (hunk 15 above)
Comment (hoopoepg) [reply src/ucp/core/ucp_ep.c]: it seems I pushed mid of changes...

File: src/ucp/core/ucp_ep.c (hunk 17 above)
Comment (hoopoepg) [reply src/ucp/core/ucp_ep.c]: added helper variable

File: src/ucp/core/ucp_ep.h (hunk 18 above)
Comment (hoopoepg) [reply src/ucp/core/ucp_ep.h]: yep, changes were not pushed...

File: src/ucp/wireup/wireup.c (hunk 20 above)
Comment (hoopoepg) [reply src/ucp/wireup/wireup.c]: ok, label is removed

File: src/ucp/wireup/wireup.c (hunk 19 above)
Comment (hoopoepg) [reply src/ucp/wireup/wireup.c:626]: partially: only scheduling of disconnect progress.
move to separate function?

Comment (yosefe) [reply src/ucp/wireup/wireup.c:626]: yes, extract common code to helper, and use different function names

**Comment by hoopoepg** [comment]
bot:pipe:retest

**Comment by hoopoepg** [comment]
@yosefe ok to squash?

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
> @yosefe ok to squash?

@hoopoepg pls squash

**Comment by hoopoepg** [comment]
squashed & rebased to master due to conflicts

**Pull request merged**